### PR TITLE
[Feat] Add config based template to query function

### PR DESCRIPTION
**Pull request by ianupamsingh**
Resolves: #170 
Changelog:
- Adds `template` as an attribute of `QueryConfig`
- Adds validation to custom `template`
- Adds `template` as an argument to `generate_prompt`

### CODE REVIEW

**Comment by cachho** [comment]
Interesting take, I would have just made the initial sentence customizable and leave the rest as it is.

File: embedchain/config/QueryConfig.py [hunk 1]
```diff
@@ -1,8 +1,35 @@
...
-    def __init__(self):
-        pass
+    def __init__(self, template: Template = None):
+        """
+        Initializes the QueryConfig instance.
+
+        :param template: Optional. The `Template` instance to use as a template for prompt.
+        :raises ValueError: If the template is not valid as template should contain $context and $query
+        """
+        template = template if template else DEFAULT_PROMPT_TEMPLATE
```
Comment (cachho) [inline embedchain/config/QueryConfig.py:31]: Can you make this 

```python
if template is None:
    template = DEFAULT_PROMPT_TEMPLATE
```

?

File: embedchain/config/QueryConfig.py [hunk 2]
```diff
@@ -1,8 +1,35 @@
...
+        if not (re.search(query_re, template.template) \
```
Comment (cachho) [inline embedchain/config/QueryConfig.py:32]: can't you use `if $query not in template.template` ?

**Comment by ianupamsingh** [comment]
Wouldn't this allow for more customization? For e.g. if I want it to reply me as a chatbot using "I" language rather than simply giving me information. In this way I can create custom characters. This is how imagined it would be used.

But if it adds more vulnerability, I can always change it. Do share your thoughts.

File: embedchain/config/QueryConfig.py (hunk 2 above)
Comment (ianupamsingh) [reply embedchain/config/QueryConfig.py:32]: Template allows both $query and ${query} formats. hence i would have to add two in operators

**Comment by cachho** [comment]
No your idea is better. I accept it.

**Comment by cachho** [comment]
Please add this to the config section in the readme. Include the variables that need to be included in the template. Since this is an important feature, and passing a template object isn't the "default behavior", please try to put an example too.

File: embedchain/config/QueryConfig.py (hunk 2 above)
Comment (cachho) [reply embedchain/config/QueryConfig.py:33]: I still think that's easier to understand and we don't have to import regex.

File: README.md [hunk 3]
```diff
@@ -315,7 +315,9 @@ This section describes all possible config options.
 
 #### **Query Config**
 
-*coming soon*
+|option|description|type|default|
+|---|---|---|---|
+|template|custom template for prompt|Template|DEFAULT_PROMPT_TEMPLATE|
```
Comment (cachho) [inline README.md:320]: Default should be the whole DEFAULT_PROMPT_TEMPLATE so people know what they're working with.

**Review by cachho — APPROVED** [review]

**Comment by cachho** [comment]
please also add the example to readme section Advanced.Configuration.Example - thanks

**Review by taranjeet — APPROVED** [review]

**Pull request merged**